xmlimplsrc/org/mozilla/javascript/xmlimpl/XmlProcessor.java

```diff
@@ -247,6 +247,11 @@ private DocumentBuilder getDocumentBuilderFromPool()
     private void returnDocumentBuilderToPool(DocumentBuilder db) {
         try {
             db.reset();
+            // DocumentBuilders are supposed to be namespace-aware.
+            // This is a sanity check for DocumentBuilder's resettability (a known bug in Android).
+            if (!db.isNamespaceAware()) {
```

Comment:
Is there somewhere else in the XML code where we set "isNamespaceAware" so that we can be sure that this test will not be triggered every time and caching will not work any more?

Reply:
Rhino gets all DocumentBuilders using DocumentBuilderFactory.newBuilder. The DocumentBuilderFactory is obtained using DocumentBuilderFactory.newInstance(). There are two places in non-test Rhino code where this newInstance() is called, and in both places newInstance() is followed by setNamespaceAware(true) 
```
        this.dom = DocumentBuilderFactory.newInstance();
        this.dom.setNamespaceAware(true);
```
setNamespaceAware(false) is never called in Rhino code.

So at the moment, all DocumentBuilders, created by Rhino, are namespace-aware. 

Is it possible that in future Rhino will also use namespace-unaware DocumentBuilders? My test shows that such builder won't be able to parse XML properly, so we cannot have namespace-unaware builders in the cache.

I thought about FEATURE-based solutions as well for this problem, like having something like FEATURE_DISABLE_DOCUMENT_BUILDER_CACHE or FEATURE_NON_RESETTABLE_DOCUMENT_BUILDER, but I picked the one above since it's less intrusive and doesn't require client changes. If you prefer FEATURE-based solution - I can upload another PR.

Reply:
Thanks for pointing that out. I think that this change is OK, then.